src/test_solvers/solvers.cu

```diff
@@ -0,0 +1,137 @@
+#include <iostream>
+
+#include "solvers.h"
+#include <cusp/monitor.h>
+#include <cusp/print.h>
+#include <cusp/krylov/gmres.h>
+
+#include <cuda_runtime.h>
+// where to perform the computation
+typedef cusp::device_memory MemorySpace;
+
+std::vector<float> cudaSolver(std::vector<int>* rowInd, std::vector<int>* colInd, std::vector<float>* val, std::vector<float>* b){
+    std::vector<float> x;
+    x.resize(b->size());
+    
+    // Set handler
+    cusolverSpHandle_t handleCusolver = NULL;
+    cusparseHandle_t handleCusparse = NULL;
+    cudaStream_t stream = NULL;
+
+    // Initialize handler
+    CUSOLVER_ERROR(cusolverSpCreate(&handleCusolver));
+    CUSPARSE_ERROR(cusparseCreate(&handleCusparse));
+    CUDA_ERROR(cudaStreamCreate(&stream));
+    CUSOLVER_ERROR(cusolverSpSetStream(handleCusolver, stream));
+    CUSPARSE_ERROR(cusparseSetStream(handleCusparse, stream));
+
+    int *d_rowInd, *d_colInd;
+    float *d_val, *d_b, *d_x;  // d_p is some mediate vector
+    int nnz = rowInd->size();     // Number of non-zero values in A
+    int m = b->size();   // Rows of the matrx A
+    float tol = 1.e-10;
+    int reorder = 0;
+    int singularity = 0;
+
+    // Allocate space on device
+    CUDA_ERROR(cudaMalloc((void**)&d_rowInd, sizeof(int)*nnz));
+    CUDA_ERROR(cudaMalloc((void**)&d_colInd, sizeof(int)*nnz));
+    CUDA_ERROR(cudaMalloc((void**)&d_val, sizeof(float)*nnz));
+    CUDA_ERROR(cudaMalloc((void**)&d_b, sizeof(float)*m));
+    CUDA_ERROR(cudaMalloc((void**)&d_x, sizeof(float)*m));
+
+    // Copy data (COO storage method)
+    CUDA_ERROR(cudaMemcpy(d_rowInd, rowInd->data(), sizeof(int)*nnz, cudaMemcpyHostToDevice));
+    CUDA_ERROR(cudaMemcpy(d_colInd, colInd->data(), sizeof(int)*nnz, cudaMemcpyHostToDevice));
+    CUDA_ERROR(cudaMemcpy(d_val, val->data(), sizeof(float)*nnz, cudaMemcpyHostToDevice));
+    CUDA_ERROR(cudaMemcpy(d_b, b->data(), sizeof(float)*m, cudaMemcpyHostToDevice));
+    CUDA_ERROR(cudaMemcpy(d_x, x.data(), sizeof(float)*m, cudaMemcpyHostToDevice));
+
+    // Create and define cusparse descriptor
+    cusparseMatDescr_t descrA = NULL;
+    CUSPARSE_ERROR(cusparseCreateMatDescr(&descrA));
+    CUSPARSE_ERROR(cusparseSetMatType(descrA, CUSPARSE_MATRIX_TYPE_GENERAL));
+    CUSPARSE_ERROR(cusparseSetMatIndexBase(descrA, CUSPARSE_INDEX_BASE_ZERO));
+
+    // COO to CSR
+    int *d_csrRowInd = NULL;
+    CUDA_ERROR(cudaMalloc((void**)&d_csrRowInd, sizeof(int)*(m+1)));   // Array length: (m+1)
+    CUSPARSE_ERROR(cusparseXcoo2csr(handleCusparse, d_rowInd, nnz, m, d_csrRowInd, CUSPARSE_INDEX_BASE_ZERO));
+
+    // QR method
+    // https://docs.nvidia.com/cuda/cusolver/index.html 
+    CUSOLVER_ERROR(cusolverSpScsrlsvqr(handleCusolver, m, nnz, descrA, d_val, d_csrRowInd, d_colInd, d_b, tol, reorder, d_x, &singularity));
+
+    // Copy data
+    CUDA_ERROR(cudaMemcpyAsync(x.data(), d_x, sizeof(float)*m, cudaMemcpyDeviceToHost, stream));
+
```

Comment:
Stuff has been `cudaMalloc`-ed above which sounds like it reserves space on the device.
Is there no need to free the memory in some way in the end ?